entity-framework/core/querying/null-semantics.md

```diff
@@ -0,0 +1,131 @@
+
+```sql
+SELECT [e].[Id], [e].[Int], [e].[NullableInt], [e].[String1], [e].[String2]
+FROM [Entities] AS [e]
+WHERE ([e].[String1] = [e].[String2]) OR ([e].[String1] IS NULL AND [e].[String2] IS NULL)
+
+SELECT [e].[Id], [e].[Int], [e].[NullableInt], [e].[String1], [e].[String2]
+FROM [Entities] AS [e]
+WHERE (([e].[String1] <> [e].[String2]) OR ([e].[String1] IS NULL OR [e].[String2] IS NULL)) AND ([e].[String1] IS NOT NULL OR [e].[String2] IS NOT NULL)
+```
+
+## Null semantics in functions
+
+Many functions in SQL can only return `null` result if some of their arguments are `null`. EF Core takes advantage of this to produce more efficient queries.
+The query below illustrates the optimization:
+
+[!code-csharp[Main](../../../samples/core/Querying/NullSemantics/Program.cs#Functions)]
+
+The generated SQL is as follows (we don't need to evaluate the `SUBSTRING` function):
+
+```sql
+SELECT [e].[Id], [e].[Int], [e].[NullableInt], [e].[String1], [e].[String2]
+FROM [Entities] AS [e]
+WHERE [e].[String1] IS NULL OR [e].[String2] IS NULL
+```
+
+The optimization can also be used for user defined functions. It can be done by adding `PropagatesNullability()` call to the relevant function parameters model configuration.
+To illustrate this, define two user functions inside the `DbContext`:
+
+[!code-csharp[Main](../../../samples/core/Querying/NullSemantics/NullSemanticsContext.cs#UdfBody)]
+
+The model configuration (inside `OnModelCreating` method) is as follows:
+
+[!code-csharp[Main](../../../samples/core/Querying/NullSemantics/NullSemanticsContext.cs#UdfModelConfiguration)]
+
+First function is configured in a standard way. The second function is configured to take advantage of the nullability propagation optimization.
+
+When issuing the following queries:
+
+[!code-csharp[Main](../../../samples/core/Querying/NullSemantics/Program.cs#UdfExamples)]
+
+We get this SQL:
+
+```sql
+SELECT [e].[Id], [e].[Int], [e].[NullableInt], [e].[String1], [e].[String2]
+FROM [Entities] AS [e]
+WHERE [dbo].[ConcatStrings]([e].[String1], [e].[String2]) IS NOT NULL
+
+SELECT [e].[Id], [e].[Int], [e].[NullableInt], [e].[String1], [e].[String2]
+FROM [Entities] AS [e]
+WHERE [e].[String1] IS NOT NULL AND [e].[String2] IS NOT NULL
+```
+
+Just like with built-in `Substring` function, the second query doesn't need to evaluate the function itself to test its nullability.
+
+> [!NOTE]
+> This optimization should only be used if the function can only return `null` becuase it's parameters are `null`.
+
+## Optimizations and considerations
+
+- Comparing non-nullable columns is simpler and faster than comparing nullable columns. Consider marking columns as non-nullable wherever it is possible.
+
+- `Equal` comparison is simpler and faster than `NotEqual`, because query doesn't need to distinguish between `null` and `false` result. Use equality comparison whenever it's possible.
```

Comment:
Should use ==/!= operators.

Reply:
Another alternative is "Checking for equality (`==`) is simpler and faster than checking for non-equality (`!=`)".

Also as above:

```suggestion
- `Equal` comparison is simpler and faster than `NotEqual`, because query doesn't need to distinguish between `null` and `false` result. Use equality comparison whenever possible.
```